### PR TITLE
MAGECLOUD-3205: Provide ElasticsSuite auto-configuration

### PR DESCRIPTION
**Pull request by shiftedreality**
### Fixed Issues (if relevant)

1. https://magento2.atlassian.net/browse/MAGECLOUD-3205

### Manual testing scenarios

1. https://jira.corp.magento.com/browse/MAGECLOUD-110

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+use Magento\MagentoCloud\Config\Environment;
+use Magento\MagentoCloud\Config\Stage\DeployInterface;
+use Magento\MagentoCloud\Package\Manager;
+
+/**
+ * Returns version of elasticsearch
+ */
+class ElasticSuite
+{
+    const NUMBER_OF_SHARDS = 3;
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php]: why do you set these default values? Default values should be common for all environment types.

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * @return array
+     */
+    private function getSearchConfig(): array
+    {
+        $esConfig = $this->environment->getRelationship('elasticsearch')[0] ?? [];
+
+        if (!isset($esConfig['host'], $esConfig['port'])) {
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php:100]: I guess we need to log this error, when customer install added elasticsuite package and he does not have installed ES. It should even fail the deploy process.

Comment (NadiyaS) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php:100]: Cover this case in src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php

**Review by NadiyaS — CHANGES_REQUESTED** [review]
Deployment was failed on the ElasticSuite installation step.
This is because installation of this module check ES connection, but at this moment there are no set configs for it. You should add necessary options to the installation CLI command *bin/magento setup:install*

**Comment by shiftedreality** [comment]
@NadiyaS it does not fail on installation for me. Will cover this case in Zephyr

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php (hunk 1 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php]: Fixed

**Comment by NadiyaS** [comment]
@shiftedreality what version of Magento, elasticsuite and ES you use?
I use Magento v.2.2.8, ElasticSuite v.2.6.8

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php [hunk 3]
```diff
@@ -85,4 +85,34 @@ public function getVersion(): string
...
+        try {
+            $response = $this->clientFactory->create()->get(sprintf(
+                '%s:%s/_template',
+                $esConfig['host'],
+                $esConfig['port']
+            ));
+            $templates = $response->getBody()->getContents();
+            $templates = json_decode($templates, true);
+
+            return $templates ? reset($templates)['settings'] : [];
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php]: hm, seems like __settings__ element is nested in __platformsh_index_settings__

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php [hunk 4]
```diff
@@ -85,4 +85,34 @@ public function getVersion(): string
 
         return $this->version;
     }
+
+    /**
+     * @return array
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php:114]: Add description to the method, and what settings are expected to be retrieved

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php (hunk 2 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php:100]: Done

**Comment by shiftedreality** [comment]
I'm using 2.7.0 with M2.3

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php [hunk 5]
```diff
@@ -56,57 +48,71 @@ class SearchEngine implements ProcessInterface
...
 
+        $this->logger->info('Updating search engine configuration.');
         $this->logger->info('Set search engine to: ' . $searchConfig['engine']);
 
-        // 2.1.x requires search config to be written to the shared config file: MAGECLOUD-1317
-        if (!$this->magentoVersion->isGreaterOrEqual('2.2')) {
-            $config = $this->sharedReader->read();
-            $config['system']['default']['catalog']['search'] = $searchConfig;
-            $this->sharedWriter->create($config);
+        try {
```
Comment (oshmyheliuk) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php]: It's not a good approach to have more than one try/catch block in one method.
You can use multiple catch block to handle different Exception.

Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php]: Good one, fixed

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php [hunk 6]
```diff
@@ -85,4 +85,34 @@ public function getVersion(): string
...
+        $relationship = $this->environment->getRelationship('elasticsearch');
+
+        if (!$relationship) {
+            return [];
+        }
+
+        $esConfig = $relationship[0];
+
+        try {
+            $response = $this->clientFactory->create()->get(sprintf(
```
Comment (oshmyheliuk) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php]: This part is similar to in `getVersion` method, you can move it to a new private method to avoid duplication.
```
 $response = $this->clientFactory->create()->get(sprintf(
                    '%s:%s',
                    $esConfig['host'],
                    $esConfig['port']
                ));
                $esConfiguration = $response->getBody()->getContents();
                $esConfiguration = json_decode($esConfiguration, true);
```

Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php]: Fixed

**Comment by oshmyheliuk** [comment]
Please add a description of a new variable into dist/.magento.env.yaml. Integration test will fail without description.

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php [hunk 7]
```diff
@@ -56,57 +47,69 @@ class SearchEngine implements ProcessInterface
...
+     * @throws ProcessException
      */
     public function execute()
     {
-        $this->logger->info('Updating search engine configuration.');
-
         $searchConfig = $this->searchEngineConfig->get();
 
+        $this->logger->info('Updating search engine configuration.');
         $this->logger->info('Set search engine to: ' . $searchConfig['engine']);
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:90]: This log will provide wrong information when ElasticSuite is installed

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php [hunk 8]
```diff
@@ -83,7 +83,7 @@ public function get(): array
     /**
      * @return array
      */
-    private function getSearchConfig()
+    private function getSearchConfig(): array
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php:94]: Add changes for ElasticSuite to make it visible in the configuration file.

Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php:94]: Please explain

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php (hunk 7 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:90]: Why?

Having this config
```
stage:
  global:
    SCD_ON_DEMAND: true
  deploy:
    ELASTIC_SUITE_CONFIGURATION:
      _merge: true
      engine:
        indices_settings:
          number_of_shards: 3
          number_of_replicas: 3
    SEARCH_CONFIGURATION:
      _merge: true
      engine: elasticsuite
```

Produces correct log

```
[2019-04-09 16:01:46] INFO: Set search engine to: elasticsuite  
```

Comment (NadiyaS) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:90]: Try case when you have nothing in SEARCH_CONFIGURATION.
BTW, the value from this config is ignored as I see, so if you have nothing or engine was set as ES, you still set elasticsuite configs if module is installed.

File: src/App/GenericException.php [hunk 9]
```diff
@@ -10,4 +10,11 @@
  */
 class GenericException extends \Exception
 {
+    /**
+     * @inheritDoc
+     */
+    public function __construct(string $message, int $code = 0, \Throwable $previous = null)
+    {
+        parent::__construct($message, $code, $previous);
+    }
```
Comment (bbatsche) [inline src/App/GenericException.php:19]: Why was this added?

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php [hunk 10]
```diff
@@ -85,4 +101,44 @@ public function getVersion(): string
...
+    {
+        $config = $this->getConfig();
+
+        if (!$config) {
+            return [];
+        }
+
+        try {
+            $templates = $this->call(sprintf(
+                '%s:%s/_template',
```
Comment (bbatsche) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php:126]: Port number should probably be `%d`

**Review by bbatsche — APPROVED** [review]
Two items, they're honestly not blockers though

File: src/App/GenericException.php (hunk 9 above)
Comment (shiftedreality) [reply src/App/GenericException.php:19]: To avoid situations when arguments has wrong order

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php (hunk 8 above)
Comment (NadiyaS) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php:94]: discussed offline

File: src/App/GenericException.php (hunk 9 above)
Comment (bbatsche) [reply src/App/GenericException.php:19]: Why would we be sending arguments to exceptions in the wrong order? And why would we allow for that? Also I don't see how this would address that.

Comment (shiftedreality) [reply src/App/GenericException.php:19]: That was so easy, like sending response code and later message, or message 2 times. Having type signature makes this explicit

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php [hunk 11]
```diff
@@ -129,6 +137,10 @@ private function getElasticSearchConfiguration(array $config)
             $engine = 'elasticsearch' . intval($esVersion);
         }
 
+        if ($this->elasticSuite->isInstalled()) {
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php]: in this case you ignore SEARCH_CONFIGURATION, isn't it?

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php [hunk 12]
```diff
@@ -129,6 +137,10 @@ private function getElasticSearchConfiguration(array $config)
...
+            $engine = 'elasticsuite';
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php]: in this case you ignore SEARCH_CONFIGURATION, isn't it?

Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php]: no, as it's merged later

File: dist/.magento.env.yaml [hunk 13]
```diff
@@ -262,6 +262,20 @@
...
+# ELASTIC_SUITE_CONFIGURATION - use this variable to retain customized ElasticSearch service settings                 #
+# between deployments.                                                                                                #
+# Magento Version: 2.1.4 and later                                                                                    #
+# Default value: not set                                                                                              #
+# Stages: deploy                                                                                                      #
+# Example:                                                                                                            #
+#          stage:                                                                                                     #
+#            deploy:                                                                                                  #
+#              ELASTIC_SUITE_CONFIGURATION:                                                                           #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: just noticed. Why elasticsuite was split with underscore? It should be one word ELASTICSUITE_CONFIGURATION

File: dist/.magento.env.yaml [hunk 14]
```diff
@@ -262,6 +262,20 @@
 #                elasticsearch_index_prefix: magento                                                                  #
 #                elasticsearch_server_timeout: '15'                                                                   #
 #######################################################################################################################
+# ELASTIC_SUITE_CONFIGURATION - use this variable to retain customized ElasticSearch service settings                 #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: please describe in what array element these configs will be set

File: dist/.magento.env.yaml [hunk 15]
```diff
@@ -262,6 +262,20 @@
...
+# between deployments.                                                                                                #
+# Magento Version: 2.1.4 and later                                                                                    #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: ElasticSuite is only for 2.2 and 2.3 Magento versions

File: dist/.magento.env.yaml (hunk 14 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Fixed

File: dist/.magento.env.yaml (hunk 15 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Thanks

File: dist/.magento.env.yaml (hunk 13 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Fixed

File: dist/.magento.env.yaml (hunk 14 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Please explain

File: dist/.magento.env.yaml [hunk 16]
```diff
@@ -262,6 +262,20 @@
...
+# Default value: not set                                                                                              #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: why not set? by default you set a lot of data if package exists. It also should be mentioned

Comment (shiftedreality) [reply dist/.magento.env.yaml]: Added

File: dist/.magento.env.yaml (hunk 14 above)
Comment (NadiyaS) [reply dist/.magento.env.yaml]: All data which customer set to this config - added to system/default/smile_elasticsuite_core_base_settings, right?
and there may be mush more configs than you added to examples, so, need to add info in what format they may add these configs. Something like: you may add/extend all configurations that are stored under path xxx for ElasticSuite settings

File: dist/.magento.env.yaml [hunk 17]
```diff
@@ -262,6 +262,20 @@
...
+# Stages: deploy                                                                                                      #
+# Example:                                                                                                            #
+#          stage:                                                                                                     #
+#            deploy:                                                                                                  #
+#              ELASTIC_SUITE_CONFIGURATION:                                                                           #
+#                engine:                                                                                              #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: are you sure that __engine__ should be here?

File: dist/.magento.env.yaml (hunk 14 above)
Comment (NadiyaS) [reply dist/.magento.env.yaml]: And add info that that engine should be changed in SEARCH_CONFIGURATION

File: dist/.magento.env.yaml (hunk 17 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Fixed

File: dist/.magento.env.yaml (hunk 14 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: Fixed

Comment (NadiyaS) [reply dist/.magento.env.yaml]: Do not see mentions about SEARCH_CONFIGURATION

File: dist/.magento.env.yaml [hunk 18]
```diff
@@ -262,6 +262,27 @@
 #                elasticsearch_index_prefix: magento                                                                  #
 #                elasticsearch_server_timeout: '15'                                                                   #
 #######################################################################################################################
+# ELASTICSUITE_CONFIGURATION - use this variable to retain customized ElasticSearch service settings                  #
+# between deployments. Is added under 'system/default/smile_elasticsuite_core_base_settings' system configuration.    #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml]: > Is added under 'system/default/smile_elasticsuite_core_base_settings' system configuration. 

> 
I guess, you wanted to add "_this config is added under_".
Or it may be rephrased to something like `All configs provided in this configuration are nested in main ElasticSuite config section __smile_elasticsuite_core_base_settings__`

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php [hunk 19]
```diff
@@ -56,57 +47,69 @@ class SearchEngine implements ProcessInterface
...
-            $config['system']['default']['catalog']['search'] = $searchConfig;
-            $this->sharedWriter->create($config);
+        $config['system']['default']['catalog']['search'] = $searchConfig;
+
+        if ($this->elasticSuite->isAvailable()) {
+            $this->logger->info('Configuring ElasticSuite');
 
-            return;
+            $config['system']['default']['smile_elasticsuite_core_base_settings'] = $this->elasticSuite->get();
+        } elseif ($this->elasticSuite->isInstalled()) {
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:98]: This validation should be moved to the beginning of the deploy phase and it should fail the deployment for now, as Magento won't work with elasticsuite and without ES, even if engine was set as __mysql__ :(

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php [hunk 20]
```diff
@@ -69,13 +87,11 @@ public function getVersion(): string
             $esConfig = $relationships['elasticsearch'][0];
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php:92]: use new methods you added getConfig() and check all other places where need to change

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+ * See COPYING.txt for license details.
+ */
+namespace Magento\MagentoCloud\Process\Deploy\InstallUpdate\ConfigUpdate\SearchEngine;
+
+use Magento\MagentoCloud\Config\ConfigMerger;
+use Magento\MagentoCloud\Config\Stage\DeployInterface;
+use Magento\MagentoCloud\Package\Manager;
+
+/**
+ * Returns version of elasticsearch
```
Comment (NadiyaS) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php]: Change the description to correct one

File: dist/.magento.env.yaml [hunk 22]
```diff
@@ -262,6 +262,27 @@
...
+# Magento Version: 2.2.0 and later                                                                                    #
```
Comment (YPyltiai) [inline dist/.magento.env.yaml:268]: Doesn't ElasticSuite work with 2.1.x?

Comment (NadiyaS) [reply dist/.magento.env.yaml:268]: According to their documentation - no
https://github.com/Smile-SA/elasticsuite/wiki/ModuleInstall#requirements

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php (hunk 19 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:98]: It is an edge case

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php (hunk 21 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php]: fixed

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php (hunk 20 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php:92]: It is out of scope of my task, I'll create tech debt

File: dist/.magento.env.yaml (hunk 18 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: fixed

File: dist/.magento.env.yaml (hunk 14 above)
Comment (shiftedreality) [reply dist/.magento.env.yaml]: We change engine automatically

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php (hunk 19 above)
Comment (NadiyaS) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:98]: We need to cover this case. Customer should know that something went wrong and he should know the reason of it. We should not allow displaying successful deploy with broken store configs.

Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:98]: If there is no ElasticSearch - default engine will be mysql. Why it's bothering you?

File: dist/.magento.env.yaml [hunk 23]
```diff
@@ -262,6 +262,28 @@
 #                elasticsearch_index_prefix: magento                                                                  #
 #                elasticsearch_server_timeout: '15'                                                                   #
 #######################################################################################################################
+# ELASTICSUITE_CONFIGURATION - use this variable to retain customized ElasticSuite service settings                   #
+# between deployments. All configs provided in this configuration are nested in main ElasticSuite                     #
+# config section 'system/default/smile_elasticsuite_core_base_settings'                                               #
+# Magento Version: 2.2.0 and later                                                                                    #
+# Default value: Depends on template configuration of ElasticSearch (i.e. default number of replicas and shards)      #
```
Comment (NadiyaS) [inline dist/.magento.env.yaml:269]: Would be good to mention _merge here. Or mention it in additional section for all such configs

Comment (shiftedreality) [reply dist/.magento.env.yaml:269]: That's general question to Docs Team

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+     * @return Validator\ResultInterface
+     */
+    public function validate(): Validator\ResultInterface
+    {
+        if (!$this->elasticSuite->isInstalled()) {
+            return $this->resultFactory->success();
+        }
+
+        if (!$this->elasticSuite->isAvailable()) {
+            return $this->resultFactory->error('ElasticSuite is installed without available ElasticSearch service.');
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: how do you know that if ElasticSuite is not available then the problem exactly in the ES ;) ?
This text of error should be generated in that class where you validate ES, or use validation of ES here.

Comment (shiftedreality) [reply src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Because of `isAvailable` method

```
$this->elasticSearch->isInstalled() && $this->isInstalled();
```

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php (hunk 7 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php:90]: Covered by validator

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php (hunk 4 above)
Comment (shiftedreality) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSearch.php:114]: Added

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php (hunk 24 above)
Comment (shiftedreality) [reply src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Added ES dependency to Validator

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        DeployInterface $config
+    ) {
+        $this->elasticSuite = $elasticSuite;
+        $this->elasticSearch = $elasticSearch;
+        $this->resultFactory = $resultFactory;
+        $this->config = $config;
+    }
+
+    /**
+     * @return Validator\ResultInterface
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Please describe all these cases for validation and why we interrupt deployment.

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+            return $this->resultFactory->success();
+        }
+
+        if (!$this->elasticSearch->isInstalled()) {
+            return $this->resultFactory->error('ElasticSuite is installed without available ElasticSearch service.');
+        }
+
+        $engine = $this->config->get(DeployInterface::VAR_SEARCH_CONFIGURATION)['engine'] ?? null;
+
+        if ($engine && strtolower($engine) !== ElasticSuite::ENGINE_NAME) {
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: I guess it may work in case of engine = elasticsearch

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+use Magento\MagentoCloud\Config\Stage\DeployInterface;
+use Magento\MagentoCloud\Config\Validator;
+use Magento\MagentoCloud\Config\ValidatorInterface;
+use Magento\MagentoCloud\Process\Deploy\InstallUpdate\ConfigUpdate\SearchEngine\ElasticSearch;
+use Magento\MagentoCloud\Process\Deploy\InstallUpdate\ConfigUpdate\SearchEngine\ElasticSuite;
+
+/**
+ * Validates different aspects of ElasticSuite's configuration.
+ */
+class ElasticSuiteConfiguration implements ValidatorInterface
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: You validate ElasticSuite itself, not its configurations. The class name is confusing a little bit

**Review by NadiyaS — COMMENTED** [review]
All left in PR comments should be fixed. 
Changes related to wrong architecture of classes src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine.php,
src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/Config.php, 
src/Process/Deploy/InstallUpdate/ConfigUpdate/SearchEngine/ElasticSuite.php may be moved to additional tech debt task or should be implemented here

**Comment by shiftedreality** [comment]
TechDebt task https://magento2.atlassian.net/browse/MAGECLOUD-3443

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php (hunk 26 above)
Comment (shiftedreality) [reply src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Wasn't working locally

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php (hunk 25 above)
Comment (shiftedreality) [reply src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Added

File: src/Config/Validator/Deploy/ElasticSuiteConfiguration.php (hunk 27 above)
Comment (shiftedreality) [reply src/Config/Validator/Deploy/ElasticSuiteConfiguration.php]: Renamed to `ElasticSuiteIntegrity`

**Review by NadiyaS — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved
